### PR TITLE
More SAML configuration options

### PR DESCRIPTION
**Pull request by robingarner**
In order to get the plugin to authenticate nicely with a Shibboleth IdP, I needed a couple of configuration items:
- an SSL key pair.  
- the ability to configure an attribute to be used as the username rather than the nameID.  
It's also very useful to be able to access the SP metadata in order to configure the IdP.  This pull request implements these changes.

The link to access the metadata is a bit clunky - it would be nice if this was a copyButton, but I couldn't see how to make its contents dynamic.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/saml/SamlEncryptionData.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package org.jenkinsci.plugins.saml;
```
Comment (0xcfff) [inline src/main/java/org/jenkinsci/plugins/saml/SamlEncryptionData.java]: All java sources in this repo contain a header with license information. I guess this practice should be applied to this file as well.

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java [hunk 2]
```diff
@@ -203,6 +225,39 @@ public HttpResponse doFinishLogin(StaplerRequest request, StaplerResponse respon
...
+    if (usernameAttributeName != null) {
+      Object attribute = saml2Profile.getAttribute(usernameAttributeName);
+      if (attribute instanceof String) {
+        return (String) attribute;
+      }
+      if (attribute instanceof List) {
+        return (String) ((List<?>)attribute).get(0);
+      }
+      LOG.log(Level.SEVERE, "Unable to get username from Saml Profile {0}", saml2Profile);
+      throw new Error("Attribute "+usernameAttributeName+" contains no usable username");
```
Comment (0xcfff) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java]: In other places of this file for similar errors IllegalStateException class is used. Should not the IllegalStateException be used here as well?

File: src/main/resources/org/jenkinsci/plugins/saml/SamlSecurityRealm/config.jelly [hunk 3]
```diff
@@ -16,5 +16,36 @@
     <f:entry title="Maximum Authentication Lifetime" field="maximumAuthenticationLifetime" help="/plugin/saml/help/maximumAuthenticationLifetime.html">
       <f:textbox default="86400" />
     </f:entry>
+    <f:entry title="Username Attribute"  field="usernameAttributeName" 
+             help="/plugin/saml/help/usernameAttributeName.html">
+      <f:textbox />
+    </f:entry>
+    
+    <f:optionalBlock title="Use Encryption ?"
```
Comment (0xcfff) [inline src/main/resources/org/jenkinsci/plugins/saml/SamlSecurityRealm/config.jelly]: The **f:entry=>table=>f:optionalBlock** chain can help here to force this block to hide optional part on initial load, see [this thread](http://jenkins-ci.361315.n4.nabble.com/How-do-I-show-other-configuration-entries-only-if-the-user-have-checked-a-checkbox-in-config-jelly-td388471.html)

File: src/main/resources/org/jenkinsci/plugins/saml/SamlSecurityRealm/config.jelly [hunk 4]
```diff
@@ -16,5 +16,36 @@
...
+      </f:entry>
+      <f:entry title="Keystore password"  field="keystorePassword" help="/plugin/saml/help/keystorePassword.html">
+        <f:textbox />
+      </f:entry>
+      <f:entry title="Private Key password"  field="privateKeyPassword" help="/plugin/saml/help/keyPassword.html">
+        <f:textbox />
+      </f:entry>
+    </f:optionalBlock>
+    
+    <!--
```
Comment (0xcfff) [inline src/main/resources/org/jenkinsci/plugins/saml/SamlSecurityRealm/config.jelly]: Looks like temporary commented out block that needs to be removed before merging to main branch.

**Comment by robingarner** [comment]
Thanks for the review, I think the last commit addresses all the issues you found.

**Comment by 0xcfff** [comment]
Ok, will look later. 
One more question. I saw you've added a URL that exposes SP metadata as XML. Is that URL accessible without authentication when SAML plugin is enabled? This actually would be extremely handy, because most of modern IdP implementations can configure RP parameters from metadata URL.

**Comment by 0xcfff** [comment]
Looks good for me. @benmccann could have a look?

**Comment by robingarner** [comment]
> I saw you've added a URL that exposes SP metadata as XML. Is that URL accessible without authentication when SAML plugin is enabled?

Yes it is - I guess since it lives in the same namespace as the finishLogin URL it inherits the same security settings.

**Comment by robingarner** [comment]
*Ping* 

Anything else I need to do to get this pull request merged ?  @benmccann ?  @0xcfff ?

Regards, Robin

**Comment by benmccann** [comment]
Sorry, I won't have time to review. Perhaps someone else can handle this one

**Comment by petehayes** [comment]
Sorry about that, pressed the wrong button!

**Comment by petehayes** [comment]
I tried running this locally after already configuring the SAML plugin and hit some NullPointers in the log when it tried to read the encryption fields while displaying the configuration panel.  

```
  public String getKeystorePath() {
    return encryptionData != null ? encryptionData.getKeystorePath() : null;
  }

  public String getKeystorePassword() {
    return encryptionData != null ? encryptionData.getKeystorePassword() : null;
  }

  public String getPrivateKeyPassword() {
    return encryptionData != null ? encryptionData.getPrivateKeyPassword() : null;
  }
```

**Comment by petehayes** [comment]
Anyone still interested in this pull request?  If not, I can work on fixing the nullpointer from above and merging it into the plugin.

**Comment by robingarner** [comment]
Just got back from vacation: I've just pushed your suggested changes.

**Comment by petehayes** [comment]
Thanks.  The other feedback I have is that when I configured it, the defaulting was such that it set the usernameAttributeName to the default value when I saved but I hadn't configured the claim to be sent back by my idP.  This blew up the username logic.  Would it be better to just log a warning if the attribute can't be found and default to the profile name?

```
  private String getUsernameFromProfile(Saml2Profile saml2Profile) {
    if (usernameAttributeName != null) {
      Object attribute = saml2Profile.getAttribute(usernameAttributeName);
      if (attribute instanceof String) {
        return (String) attribute;
      }
      if (attribute instanceof List) {
        return (String) ((List<?>)attribute).get(0);
      }
      LOG.log(Level.WARNING, "Unable to get username from Saml Profile {0}, using profile ID instead", saml2Profile);
    }

    return saml2Profile.getId();
  }
```

**Comment by petehayes** [comment]
Hmm, that might not be enough actually.  If I clear out the display name claim, and save it.  When I re-open the security configuration it reappears.

**Comment by petehayes** [comment]
Sorry, that was display name not user name.  User name field doesn't automatically populate. Still I receive that error even when usernameAttributeName is left blank.  It must not be getting it as a null or I'm doing something wrong.

**Comment by robingarner** [comment]
Your approach seems sound to me.  In the constructor,
`this.usernameAttributeName = Util.fixEmptyAndTrim(usernameAttributeName);`
should convert blanks to null, so I've no idea why it's not working.

I've made the change you suggested, and it seems to work for me.  In my testing it correctly translates an empty `usernameAttributeName` to `null`, so not sure why you're seeing different.

I've given you collaborator status on my fork, so feel free to push fixes.

**Comment by petehayes** [comment]
Not sure really what happened but your change looks good to me. I will merge in the pull request.

**Comment by petehayes** [comment]
I tried to run the plugin release but it fails during the git push.  I will play with it some more.

**Pull request merged**